### ISSUE TITLE
Poor horizontal spacing - Basic combinations of durations are very uneven

### ISSUE THREAD

**Issue by Silverwolf90** [0xfe/vexflow#416]
Examples:

![image](https://cloud.githubusercontent.com/assets/1631625/17650755/78139326-624d-11e6-9b3f-8fcb7b33e71f.png)

![image](https://cloud.githubusercontent.com/assets/1631625/17650757/8781afbe-624d-11e6-8534-f24d08ba81f2.png)

![image](https://cloud.githubusercontent.com/assets/1631625/17650729/b68ef3c6-624c-11e6-9658-4c1d4af2c36d.png)

**Comment by Silverwolf90** [comment]
Worth looking at this:

"Algorithms and Data Structures for a Music Notation System based on GUIDO Music Notation"
Specifically the section starting on page 85: "Spacing, Line Breaking, and Page Filling'
http://tuprints.ulb.tu-darmstadt.de/epda/000265/kai_renz_diss.pdf

**Comment by 0xfe** [comment]
Ugh... not sure why that's happening. I'll look into it.

**Comment by 0xfe** [comment]
Okay, I think I know what's going on.

**Comment by 0xfe** [comment]
I've got a better formatter coming -- doesn't fix this specific issue, but does fix a lot of other ones. Also, it's much simpler.

**Comment by 0xfe** [comment]
I pushed a new formatter that works much better, but there's still a fundamental issue with symmetry when you have to align voices that have overlapping odd rhythms. I have two ideas for algorithms to deal with this -- one is a simple iterative technique, but slow, and the other is quite complex, but fast. :-/

I need to think this through a bit.

**Comment by Silverwolf90** [comment]
It's quite a hard problem.

I think lilypond uses a constraint solver for horizontal spacing.

Some PDF's are available that describe algorithms:

"An Improved Algorithm for Spacing a Line of Music"
http://quod.lib.umich.edu/cgi/p/pod/dod-idx/improved-algorithm-for-spacing-a-line-of-music.pdf?c=icmc;idno=bbp2372.2002.097

"A New Algorithm for Horizontal Spacing of Printed Music"
http://quod.lib.umich.edu/cgi/p/pod/dod-idx/new-algorithm-for-horizontal-spacing-of-printed-music.pdf?c=icmc;idno=bbp2372.1995.032

Taken from earlier, same author as previous link:
"Algorithms and Data Structures for a Music Notation System based on GUIDO Music Notation"
Specifically the section starting on page 85: "Spacing, Line Breaking, and Page Filling'
http://tuprints.ulb.tu-darmstadt.de/epda/000265/kai_renz_diss.pdf

**Comment by 0xfe** [comment]
Thanks, I was just reading the section from the first PDF you sent (thanks!) I noticed that the current spacing algorithm performs a little like the Sibelius example.

<img width="539" alt="screen shot 2016-08-14 at 7 51 20 pm" src="https://cloud.githubusercontent.com/assets/241299/17652954/844fb5b0-6258-11e6-8142-e18db6e84d84.png">


Vexflow:

<img width="585" alt="screen shot 2016-08-14 at 7 49 58 pm" src="https://cloud.githubusercontent.com/assets/241299/17652949/5f4ccc76-6258-11e6-9f0d-012d97a09c72.png">

**Comment by 0xfe** [comment]
I added some code to export and plot formatter decisions, in particular the gaps between ticks, which I'm going to use this to build the iterative algorithm. This specific test has about 82 pixels of freedom in aggregate, and we should be able to redistribute those pixels to format the notation more evenly.

<img width="565" alt="screen shot 2016-08-15 at 9 46 55 am" src="https://cloud.githubusercontent.com/assets/241299/17666300/a978b312-62cd-11e6-9d4b-d8c8b91a84ca.png">

**Comment by 0xfe** [comment]
I got the iterative formatter working. It's not perfect, but it's much better:

### Before
<img width="550" alt="screen shot 2016-08-15 at 2 23 41 pm" src="https://cloud.githubusercontent.com/assets/241299/17674737/cacb57fa-62f4-11e6-9358-349696d4c208.png">

### After
<img width="562" alt="screen shot 2016-08-15 at 2 23 19 pm" src="https://cloud.githubusercontent.com/assets/241299/17674743/d4f8e134-62f4-11e6-8bff-5cacfe4e0e26.png">

### Progression
Here you can see how the loss is reduced through each iteration. (Higher loss = uglier.)

<img width="536" alt="screen shot 2016-08-15 at 2 24 14 pm" src="https://cloud.githubusercontent.com/assets/241299/17674798/0d75aaf6-62f5-11e6-85ad-ca0c007d682a.png">

**Comment by Silverwolf90** [comment]
Awesome! That's way better!

**Comment by 0xfe** [comment]
Latest update is even better, and I have a pretty good idea for dealing with the fine tuning. But there are some cases where the formatter makes things worse, and I'm looking into that.

<img width="551" alt="screen shot 2016-08-15 at 4 20 44 pm" src="https://cloud.githubusercontent.com/assets/241299/17678296/87dce174-6304-11e6-806b-536573220494.png">

<img width="562" alt="screen shot 2016-08-15 at 4 38 11 pm" src="https://cloud.githubusercontent.com/assets/241299/17678847/ce420796-6306-11e6-8879-d650448a2589.png">

**Comment by mscuthbert** [comment]
This looks very good from my perspective. Congrats!

**Comment by rvilarl** [comment]
@0xfe I think that this one can be closed
